### DISCUSSION TITLE
[Winui3] Drag and Dop funtion, how to fill DataPackage with https resource ?

### DISCUSSION THREAD

**Discussion by TanWei** [microsoft/microsoft-ui-xaml#7544] [Q&A]
I've written a little file-transfer application written in C++ using WinUI ... it logs into a http server, shows the user a list of files available on the server, and lets the user upload or download files.

Now I have a request that my users would like to be able to drag a file out of the server-files-list-view and onto the desktop, or into an open folder window when i know the resource https path.

As we can see in https://docs.microsoft.com/en-us/windows/apps/design/input/drag-and-drop

I know how to fill data package object with local file, but I want to fill a data package object with https async.

In short, there are a Drag Source(my app) and a Drag Target(folder, Desktop and other app).

Drag the control of Drag Source(my app) to Drag Target(folder, Desktop and other app), trigger https download, download the file to the Drag Target.

The difficulty lies in how to download https resources when the Drag Target responds to the “Drop” Event.

more information:
I know that for the drag and drop function in win32 API (https://docs.microsoft.com/en-us/windows/win32/api/ole2/nf-ole2-dodragdrop), you can implement IDataObject yourself, in the IDataObject::GetData interface perform a download operation. Is there a way to achieve the same in winui？

@castorix Do you have any good idea?

Code show as below:
~~~
 // DragStarting Event
 void winrt::drag5::implementation::MainPage::GridViewItemPresenter_DragStarting(winrt::Microsoft::UI::Xaml::UIElement const& sender,
         winrt::Microsoft::UI::Xaml::DragStartingEventArgs const& e) {
     e.Data().RequestedOperation(DataPackageOperation::Copy);
     e.Data().SetDataProvider(StandardDataFormats::StorageItems(),
     [this](auto&& request) {
         DelyRenderer(request);
     });
     e.Data().OperationCompleted(
     [this](DataPackage data, OperationCompletedEventArgs const & args) {
     }
     );
 }
    
 fire_and_forget winrt::drag5::implementation::MainPage::DelyRenderer(DataProviderRequest request) {
     auto deferral = request.GetDeferral();
     try {
    
         // L"http://***.zip"
         // for example:https://codeload.github.com/microsoft/vscode/zip/refs/tags/1.69.2
         // how can i put the hypelink int to data
         request.SetData(data);
     } catch (...) {
    
     }
     deferral.Complete();
     co_return;
 }
~~~

**Comment by gautambjain** [comment]
You can fill the https url in the data package and use the url during the drop event. Or am I missing something?

**Reply by TanWei** [reply]
My drop target is desktop, so the drop event is handled by explorer.exe, not me
